File: pedal-beu-back/internal/services/restaurant.go
package services

import (
	"context"
	"errors"
	"log"
	"time"

	"github.com/haile-paa/pedal-delivery/internal/models"
	"github.com/haile-paa/pedal-delivery/internal/repositories"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type RestaurantService interface {
	CreateRestaurant(ctx context.Context, ownerID string, req models.CreateRestaurantRequest) (*models.Restaurant, error)
	GetRestaurantByID(ctx context.Context, id string) (*models.Restaurant, error)
	GetRestaurants(ctx context.Context, query models.RestaurantQuery) ([]models.Restaurant, int64, error)
	FindNearby(ctx context.Context, location models.GeoLocation, radius float64) ([]models.Restaurant, error)
	Search(ctx context.Context, query string, location *models.GeoLocation) ([]models.Restaurant, error)
	GetMenuItems(ctx context.Context, restaurantID string) ([]models.MenuItem, error)
	UpdateRestaurant(ctx context.Context, id string, update models.UpdateRestaurantRequest) (*models.Restaurant, error)
	DeleteRestaurant(ctx context.Context, id string) error
	AddMenuItem(ctx context.Context, restaurantID string, req models.CreateMenuItemRequest) (*models.MenuItem, error)
	UpdateMenuItem(ctx context.Context, restaurantID, itemID string, req models.UpdateMenuItemRequest) (*models.MenuItem, error)
}

type restaurantService struct {
	repo repositories.RestaurantRepository
}

func NewRestaurantService(repo repositories.RestaurantRepository) RestaurantService {
	return &restaurantService{repo: repo}
}

func (s *restaurantService) CreateRestaurant(ctx context.Context, ownerID string, req models.CreateRestaurantRequest) (*models.Restaurant, error) {
	ownerObjectID, err := primitive.ObjectIDFromHex(ownerID)
	if err != nil {
		return nil, errors.New("invalid owner ID")
	}

	// Convert menu items if provided
	var menuItems []models.MenuItem
	if req.Menu != nil {
		for _, itemReq := range req.Menu {
			// Default IsAvailable to true if not provided
			isAvailable := true
			if itemReq.IsAvailable != nil {
				isAvailable = *itemReq.IsAvailable
			}

			menuItem := models.MenuItem{
				ID:              primitive.NewObjectID(),
				Name:            itemReq.Name,
				Description:     itemReq.Description,
				Price:           itemReq.Price,
				Category:        itemReq.Category,
				Ingredients:     itemReq.Ingredients,
				Addons:          itemReq.Addons,
				IsAvailable:     isAvailable,
				PreparationTime: itemReq.PreparationTime,
				Image:           itemReq.Image,
				CreatedAt:       time.Now(),
				UpdatedAt:       time.Now(),
			}
			menuItems = append(menuItems, menuItem)
		}
	}

	restaurant := &models.Restaurant{
		OwnerID:      ownerObjectID,
		Name:         req.Name,
		Description:  req.Description,
		CuisineType:  req.CuisineType,
		Address:      req.Address,
		Phone:        req.Phone,
		Email:        req.Email,
		DeliveryFee:  req.DeliveryFee,
		MinOrder:     req.MinOrder,
		DeliveryTime: req.DeliveryTime,
		Location: models.GeoLocation{
			Type:        "Point",
			Coordinates: []float64{req.Longitude, req.Latitude},
		},
		Images:       req.Images,
		Menu:         menuItems,
		OpeningHours: models.OpeningHours{},
	}

	if err := s.repo.Create(ctx, restaurant); err != nil {
		return nil, err
	}

	return restaurant, nil
}
func (s *restaurantService) GetRestaurantByID(ctx context.Context, id string) (*models.Restaurant, error) {
	objectID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		return nil, errors.New("invalid restaurant ID")
	}

	return s.repo.FindByID(ctx, objectID)
}

func (s *restaurantService) GetRestaurants(ctx context.Context, query models.RestaurantQuery) ([]models.Restaurant, int64, error) {
	pagination := repositories.Pagination{
		Page:    int64(query.Page),
		Limit:   int64(query.Limit),
		SortBy:  "created_at",
		SortDir: -1,
	}

	log.Printf("🔍 Service: GetRestaurants called with query: %+v", query)

	// If location is provided, use FindNearby
	if query.Latitude != 0 && query.Longitude != 0 {
		log.Printf("📍 Using location-based search: lat=%f, lng=%f", query.Latitude, query.Longitude)
		location := models.GeoLocation{
			Type:        "Point",
			Coordinates: []float64{query.Longitude, query.Latitude},
		}
		radius := query.Radius
		if radius == 0 {
			radius = 10000 // Default 10km
		}
		log.Printf("📍 Search radius: %.0f meters", radius)

		// Call repository and handle response
		restaurants, total, err := s.repo.FindNearby(ctx, location, radius, pagination)
		if err != nil {
			log.Printf("❌ Error in FindNearby: %v", err)
			return []models.Restaurant{}, 0, err
		}

		log.Printf("✅ FindNearby returned %d restaurants, total: %d", len(restaurants), total)
		if len(restaurants) == 0 {
			log.Printf("⚠️ No restaurants found near the location")
			return []models.Restaurant{}, 0, nil // Return empty slice, not nil
		}

		return restaurants, total, nil
	}

	// If no location, get all restaurants without geospatial query
	log.Printf("📍 No location provided, getting all restaurants")
	restaurants, total, err := s.repo.FindAll(ctx, pagination)
	if err != nil {
		log.Printf("❌ Error in FindAll: %v", err)
		return []models.Restaurant{}, 0, err
	}

	log.Printf("✅ FindAll returned %d restaurants, total: %d", len(restaurants), total)
	if len(restaurants) == 0 {
		log.Printf("⚠️ No restaurants found in database")
		return []models.Restaurant{}, 0, nil // Return empty slice, not nil
	}

	return restaurants, total, nil
}

func (s *restaurantService) FindNearby(ctx context.Context, location models.GeoLocation, radius float64) ([]models.Restaurant, error) {
	pagination := repositories.Pagination{
		Page:    1,
		Limit:   50,
		SortBy:  "rating",
		SortDir: -1,
	}

	restaurants, _, err := s.repo.FindNearby(ctx, location, radius, pagination)
	return restaurants, err
}

func (s *restaurantService) Search(ctx context.Context, query string, location *models.GeoLocation) ([]models.Restaurant, error) {
	radius := 10000.0 // 10km default radius
	if location == nil {
		// Search without location - use a default location
		defaultLocation := models.GeoLocation{
			Type:        "Point",
			Coordinates: []float64{38.746, 9.032}, // Addis Ababa coordinates
		}
		location = &defaultLocation
	}

	return s.repo.Search(ctx, query, *location, radius)
}

func (s *restaurantService) GetMenuItems(ctx context.Context, restaurantID string) ([]models.MenuItem, error) {
	objectID, err := primitive.ObjectIDFromHex(restaurantID)
	if err != nil {
		return nil, errors.New("invalid restaurant ID")
	}

	return s.repo.GetMenuItems(ctx, objectID)
}

func (s *restaurantService) UpdateRestaurant(ctx context.Context, id string, req models.UpdateRestaurantRequest) (*models.Restaurant, error) {
	objectID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		return nil, errors.New("invalid restaurant ID")
	}

	update := bson.M{}
	if req.Name != "" {
		update["name"] = req.Name
	}
	if req.Description != "" {
		update["description"] = req.Description
	}
	if len(req.CuisineType) > 0 {
		update["cuisine_type"] = req.CuisineType
	}
	if req.Address != "" {
		update["address"] = req.Address
	}
	if req.Phone != "" {
		update["phone"] = req.Phone
	}
	if req.Email != "" {
		update["email"] = req.Email
	}
	if req.DeliveryFee > 0 {
		update["delivery_fee"] = req.DeliveryFee
	}
	if req.MinOrder > 0 {
		update["min_order"] = req.MinOrder
	}
	if req.DeliveryTime > 0 {
		update["delivery_time"] = req.DeliveryTime
	}
	if req.Latitude != 0 && req.Longitude != 0 {
		update["location"] = models.GeoLocation{
			Type:        "Point",
			Coordinates: []float64{req.Longitude, req.Latitude},
		}
	}

	// Handle Images field
	if req.Images != nil {
		update["images"] = req.Images
	}

	// Handle Menu field if provided
	if req.Menu != nil {
		var menuItems []models.MenuItem
		for _, itemReq := range req.Menu {
			// Default IsAvailable to true if not provided
			isAvailable := true
			if itemReq.IsAvailable != nil {
				isAvailable = *itemReq.IsAvailable
			}

			menuItem := models.MenuItem{
				ID:              primitive.NewObjectID(),
				Name:            itemReq.Name,
				Description:     itemReq.Description,
				Price:           itemReq.Price,
				Category:        itemReq.Category,
				Ingredients:     itemReq.Ingredients,
				Addons:          itemReq.Addons,
				IsAvailable:     isAvailable,
				PreparationTime: itemReq.PreparationTime,
				Image:           itemReq.Image,
				CreatedAt:       time.Now(),
				UpdatedAt:       time.Now(),
			}
			menuItems = append(menuItems, menuItem)
		}
		update["menu"] = menuItems
	}

	if err := s.repo.Update(ctx, objectID, update); err != nil {
		return nil, err
	}

	return s.repo.FindByID(ctx, objectID)
}
func (s *restaurantService) AddMenuItem(ctx context.Context, restaurantID string, req models.CreateMenuItemRequest) (*models.MenuItem, error) {
	objectID, err := primitive.ObjectIDFromHex(restaurantID)
	if err != nil {
		return nil, errors.New("invalid restaurant ID")
	}

	menuItem := &models.MenuItem{
		ID:              primitive.NewObjectID(),
		Name:            req.Name,
		Description:     req.Description,
		Price:           req.Price,
		Category:        req.Category,
		Ingredients:     req.Ingredients,
		Addons:          req.Addons,
		IsAvailable:     true,
		PreparationTime: req.PreparationTime,
		Image:           req.Image,
		CreatedAt:       time.Now(),
		UpdatedAt:       time.Now(),
	}

	if err := s.repo.AddMenuItem(ctx, objectID, menuItem); err != nil {
		return nil, err
	}

	return menuItem, nil
}

func (s *restaurantService) DeleteRestaurant(ctx context.Context, id string) error {
	objectID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		return errors.New("invalid restaurant ID")
	}

	// We need to add a Delete method to your repository
	// For now, let's update the status to inactive
	update := bson.M{
		"is_active": false,
	}
	return s.repo.Update(ctx, objectID, update)
}

func (s *restaurantService) UpdateMenuItem(ctx context.Context, restaurantID, itemID string, req models.UpdateMenuItemRequest) (*models.MenuItem, error) {
	restaurantObjectID, err := primitive.ObjectIDFromHex(restaurantID)
	if err != nil {
		return nil, errors.New("invalid restaurant ID")
	}

	itemObjectID, err := primitive.ObjectIDFromHex(itemID)
	if err != nil {
		return nil, errors.New("invalid menu item ID")
	}

	update := bson.M{}
	if req.Name != "" {
		update["name"] = req.Name
	}
	if req.Description != "" {
		update["description"] = req.Description
	}
	if req.Price > 0 {
		update["price"] = req.Price
	}
	if req.Category != "" {
		update["category"] = req.Category
	}
	if req.Ingredients != nil {
		update["ingredients"] = req.Ingredients
	}
	if req.Addons != nil {
		update["addons"] = req.Addons
	}
	if req.IsAvailable != nil {
		update["is_available"] = *req.IsAvailable
	}
	if req.PreparationTime > 0 {
		update["preparation_time"] = req.PreparationTime
	}
	if req.Image != "" {
		update["image"] = req.Image
	}

	if err := s.repo.UpdateMenuItem(ctx, restaurantObjectID, itemObjectID, update); err != nil {
		return nil, err
	}

	// Fetch and return the updated menu item
	// We need to get the restaurant and find the menu item
	restaurant, err := s.repo.FindByID(ctx, restaurantObjectID)
	if err != nil {
		return nil, err
	}

	for _, item := range restaurant.Menu {
		if item.ID == itemObjectID {
			return &item, nil
		}
	}

	return nil, errors.New("menu item not found")
}

File: pedal-beu-front/lib/restaurant.ts
// lib/restaurant.ts
import api from "./api";
import { Restaurant, MenuItem, Addon } from "../src/types";

// Retry function for network requests
const retryRequest = async <T>(
  requestFn: () => Promise<T>,
  maxRetries: number = 3,
  delayMs: number = 1000,
): Promise<T> => {
  let lastError: any;

  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      return await requestFn();
    } catch (error: any) {
      lastError = error;

      // Only retry on network/timeout errors
      const shouldRetry =
        error.code === "ECONNABORTED" ||
        error.message === "Network Error" ||
        error.code === "ERR_NETWORK";

      if (!shouldRetry || attempt === maxRetries - 1) {
        throw error;
      }

      console.log(
        `Retry attempt ${attempt + 1}/${maxRetries} after error:`,
        error.message,
      );
      await new Promise((resolve) =>
        setTimeout(resolve, delayMs * (attempt + 1)),
      );
    }
  }

  throw lastError;
};

export const restaurantAPI = {
  // lib/restaurant.ts
  getAll: async (params?: any) => {
    try {
      console.log("restaurantAPI.getAll called with params:", params);

      const queryParams: any = {
        page: params?.page || 1,
        limit: params?.limit || 20,
      };

      if (params?.category) {
        queryParams.category = params.category;
      }

      if (params?.location) {
        queryParams.latitude = params.location.latitude;
        queryParams.longitude = params.location.longitude;
        params.radius = params.location.radius || 10000;
        queryParams.calculate_distance = true;
      }

      console.log(
        "Making API request to /restaurants with params:",
        queryParams,
      );

      // Wrap the API call with retry logic
      const response = await retryRequest(
        () =>
          api.get("/restaurants", {
            params: queryParams,
            timeout: 30000, // Increased from 10000 to 30000
          }),
        3, // max retries
        1000, // initial delay
      );

      console.log("API Response received:", {
        status: response.status,
        statusText: response.statusText,
        data: response.data,
      });

      // Handle the response structure
      const responseData = response.data;

      // Check if data is null or undefined
      if (!responseData || !responseData.data) {
        console.log("No data in response, returning empty array");
        return {
          success: true,
          data: [],
          pagination: responseData?.pagination || {
            page: 1,
            limit: 20,
            total: 0,
            total_pages: 0,
          },
        };
      }

      // Data might be null or an array
      const rawData = responseData.data;
      let restaurants: Restaurant[] = [];

      if (Array.isArray(rawData)) {
        console.log(`Received ${rawData.length} restaurants`);

        // Map the data
        restaurants = rawData.map((item: any) => {
          // If the item has a restaurant property (when calculate_distance is true)
          if (item.restaurant) {
            return {
              id: item.restaurant.id || item.restaurant._id || "",
              owner_id: item.restaurant.owner_id || "",
              name: item.restaurant.name || "",
              description: item.restaurant.description || "",
              cuisine_type: item.restaurant.cuisine_type || [],
              location: item.restaurant.location || {
                type: "Point",
                coordinates: [0, 0],
              },
              address: item.restaurant.address || "",
              phone: item.restaurant.phone || "",
              email: item.restaurant.email || "",
              images: item.restaurant.images || [],
              menu: item.restaurant.menu || [],
              is_active: item.restaurant.is_active ?? true,
              is_verified: item.restaurant.is_verified ?? false,
              rating: item.restaurant.rating || 0,
              total_reviews: item.restaurant.total_reviews || 0,
              opening_hours: item.restaurant.opening_hours || {},
              delivery_fee: item.restaurant.delivery_fee || 0,
              min_order: item.restaurant.min_order || 0,
              delivery_time: item.restaurant.delivery_time || 30,
              created_at:
                item.restaurant.created_at || new Date().toISOString(),
              updated_at:
                item.restaurant.updated_at || new Date().toISOString(),
              distance_km: item.distance_km,
              distance_m: item.distance_m,
            };
          } else {
            // Regular restaurant object
            return {
              id: item.id || item._id || "",
              owner_id: item.owner_id || "",
              name: item.name || "",
              description: item.description || "",
              cuisine_type: item.cuisine_type || [],
              location: item.location || {
                type: "Point",
                coordinates: [0, 0],
              },
              address: item.address || "",
              phone: item.phone || "",
              email: item.email || "",
              images: item.images || [],
              menu: item.menu || [],
              is_active: item.is_active ?? true,
              is_verified: item.is_verified ?? false,
              rating: item.rating || 0,
              total_reviews: item.total_reviews || 0,
              opening_hours: item.opening_hours || {},
              delivery_fee: item.delivery_fee || 0,
              min_order: item.min_order || 0,
              delivery_time: item.delivery_time || 30,
              created_at: item.created_at || new Date().toISOString(),
              updated_at: item.updated_at || new Date().toISOString(),
            };
          }
        });
      } else {
        console.log("Response data is not an array:", rawData);
      }

      console.log(`Mapped ${restaurants.length} restaurants`);

      return {
        success: true,
        data: restaurants,
        pagination: responseData.pagination || {
          page: 1,
          limit: 20,
          total: 0,
          total_pages: 0,
        },
      };
    } catch (error: any) {
      // Don't log as error if it's a network/timeout error
      if (
        error.code === "ECONNABORTED" ||
        error.message === "Network Error" ||
        error.code === "ERR_NETWORK"
      ) {
        console.log(
          "Network/timeout error in restaurantAPI.getAll after retries:",
          {
            message: error.message,
            code: error.code,
          },
        );
      } else {
        console.error("Error in restaurantAPI.getAll:", error);
      }

      return {
        success: false,
        data: [],
        error:
          error.response?.data?.error ||
          error.message ||
          "Failed to fetch restaurants",
      };
    }
  },

  // Get restaurant by ID with retry
  getById: async (id: string) => {
    try {
      console.log(`Fetching restaurant ${id}...`);
      const response = await retryRequest(
        () => api.get(`/restaurants/${id}`),
        2, // max retries
        500, // initial delay
      );
      console.log(`Restaurant ${id} response:`, response.data);
      return { success: true, data: response.data };
    } catch (error: any) {
      console.log(`Error fetching restaurant ${id}:`, error.message);
      return {
        success: false,
        error: error.response?.data?.error || "Failed to fetch restaurant",
      };
    }
  },

  // Get nearby restaurants with retry
  getNearby: async (
    latitude: number,
    longitude: number,
    radius: number = 10000, // Default 10km to match backend
  ): Promise<{ success: boolean; data: Restaurant[]; error?: string }> => {
    try {
      const response = await retryRequest(
        () =>
          api.get("/restaurants/nearby", {
            params: { latitude, longitude, radius },
          }),
        2,
        500,
      );
      return { success: true, data: response.data.data || response.data };
    } catch (error: any) {
      return {
        success: false,
        data: [],
        error:
          error.response?.data?.error || "Failed to fetch nearby restaurants",
      };
    }
  },

  // Search restaurants with retry
  search: async (
    query: string,
    location?: { latitude: number; longitude: number },
  ): Promise<{ success: boolean; data: Restaurant[]; error?: string }> => {
    try {
      const params: any = { q: query };
      if (location) {
        params.latitude = location.latitude;
        params.longitude = location.longitude;
      }
      const response = await retryRequest(
        () => api.get("/restaurants/search", { params }),
        2,
        500,
      );
      return { success: true, data: response.data.data || response.data };
    } catch (error: any) {
      return {
        success: false,
        data: [],
        error: error.response?.data?.error || "Failed to search restaurants",
      };
    }
  },

  // Get menu items for restaurant with retry
  getMenu: async (
    restaurantId: string,
  ): Promise<{ success: boolean; data: MenuItem[]; error?: string }> => {
    try {
      const response = await retryRequest(
        () => api.get(`/restaurants/${restaurantId}/menu`),
        2,
        500,
      );
      return { success: true, data: response.data };
    } catch (error: any) {
      return {
        success: false,
        data: [],
        error: error.response?.data?.error || "Failed to fetch menu",
      };
    }
  },

  // ... rest of the functions remain the same (Admin functions)
  createRestaurant: async (data: {
    name: string;
    description: string;
    cuisine_type: string[];
    address: string;
    latitude: number;
    longitude: number;
    phone: string;
    email?: string;
    delivery_fee: number;
    min_order: number;
    delivery_time: number;
  }): Promise<{ success: boolean; data?: Restaurant; error?: string }> => {
    try {
      const response = await retryRequest(
        () => api.post("/restaurants", data),
        2,
        500,
      );
      return {
        success: true,
        data: response.data.data || response.data,
      };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.error || "Failed to create restaurant",
      };
    }
  },

  updateRestaurant: async (
    id: string,
    data: Partial<Restaurant>,
  ): Promise<{ success: boolean; data?: Restaurant; error?: string }> => {
    try {
      const response = await retryRequest(
        () => api.put(`/restaurants/${id}`, data),
        2,
        500,
      );
      return { success: true, data: response.data.data || response.data };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.error || "Failed to update restaurant",
      };
    }
  },

  addMenuItem: async (
    restaurantId: string,
    data: {
      name: string;
      description: string;
      price: number;
      category: string;
      ingredients: string[];
      addons?: Addon[];
      preparation_time: number;
      image?: string;
    },
  ): Promise<{ success: boolean; data?: MenuItem; error?: string }> => {
    try {
      const response = await retryRequest(
        () => api.post(`/restaurants/${restaurantId}/menu`, data),
        2,
        500,
      );
      return { success: true, data: response.data.data || response.data };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.error || "Failed to add menu item",
      };
    }
  },

  updateMenuItem: async (
    restaurantId: string,
    itemId: string,
    data: Partial<MenuItem>,
  ): Promise<{ success: boolean; data?: MenuItem; error?: string }> => {
    try {
      const response = await retryRequest(
        () => api.put(`/restaurants/${restaurantId}/menu/${itemId}`, data),
        2,
        500,
      );
      return { success: true, data: response.data.data || response.data };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.error || "Failed to update menu item",
      };
    }
  },

  deleteMenuItem: async (
    restaurantId: string,
    itemId: string,
  ): Promise<{ success: boolean; error?: string }> => {
    try {
      await retryRequest(
        () => api.delete(`/restaurants/${restaurantId}/menu/${itemId}`),
        2,
        500,
      );
      return { success: true };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.error || "Failed to delete menu item",
      };
    }
  },
};
